content/manager/high-availability-clusters.md

```diff
@@ -0,0 +1 @@
draft: true
weight: 899

---


#Using Clusters to Provide High Availability

If you have an Enterprise version of Cloudify Manager, an administrator user can create a cluster of up to three Cloudify Managers to enable high availability. One Cloudify Manager is designated as the master (the active Cloudify Manager) and the others are designated as hot standbys that are constant mirrors of the data of the master. In the event that the master Cloudify Manager fails, an automatic failover switch activates one of the hot standbys as the master.

The cluster is dynamic, meaning that you do not need to specify the size of the cluster in advance. You just specify `cluster-join` when you bring the Cloudify Manager up. To ensure maximal performance, it is recommended that you do not have more than three Cloudify Managers in a cluster.

##Creating a Cluster
You create a cluster at the time that you are bootstrapping your Cloudify Managers. During configuration, high availability is configured automatically when you run the `cluster-start` command on the first Cloudify Manager in the cluster. The additional Cloudify Managers use the `cluster-join` command during bootstrapping.

The Cloudify Managers that you join to the cluster must be in an empty state, otherwise existing data will be lost. The state of each Cloudify Manager is linked to the state of the cluster. Operations can only be performed on the master in the cluster, but are reflected on all Cloudify Managers. Similarly, upload requests can only be sent to the master.

You do not need to know which of the Cloudify Managers is the master, the console and console agent automatically detect which one is active, whether Cloudify Manager, the CLI or the Web interface are being used. However, the you do need to know which Cloudify Manager nodes are in the cluster.

##Failure of the Master Cloudify Manager
As already mentioned, in the event that the active Cloudify Manager fails, one of the hot standbys is activated. It is important that the user investigates and fixes the issues that caused the original master to fail so that high availability is maintained and to avoid having a single point of failure.

##Important Considerations
Be aware of the following considerations when you are implementing clustering.

* If the active Cloudify Manager was processing a workflow at the time it fails, the new active manager does not resume and complete the workflow.
* Policies are not synchronized between Cloudify Managers in the cluster.
* If there is a loss of connection between the Cloudify Managers in the cluster, they independently assume the `active` role. When the connection is resumed, the most recently updated Cloudify Manager becomes the master. Data that was accumulated on the other Cloudify Managers during the disconnection is not synchronized, so is lost. 
* A Cloudify Manager that is down remains in the cluster unless you remove it. To remove a Cloudify Manager, run `cluster-remove-node`.



```

Comment:
1. Enterprise version -> rename to Premium version
2. I'd remove the first disclaimer about 3 managers in a cluster, as it is described in the details section
3. You just specify `cluster-join` when you bring the Cloudify Manager up - you first specify cluster-start to start a cluster, then cluster-join from any manager that joins the cluster - so I would either replace the Cloudify manager with "a Cloudify manager" or add this clarification
4. Creating a cluster - can be done at any time post bootstrap, even after there is data in the manager
5. cluster-join - command should be triggered post bootstrap. The joining manager has to be empty or the command fails
6. State is linked - worth noting this is the data kept by the manager, not the manager's state (running, stopped, etc)
7. I'd change the sentence with console to "Cloudify automatically detects the active manager, using Console"
8. "However, the you do need to know which Cloudify Manager nodes are in the cluster" - not sure I understand this one... You don't actually, as a user, have to know this :)
9. "It is important that the user investigates and fixes the issues that caused the original master to fail" - or alternatively add another manager to the cluster. Note that since operations cannot be performed on the non active manager, troubleshooting has to be done by SSHing the manager.